python/lib/communication/dmod/communication/maas_request/model_exec_request_response.py

```diff
@@ -1,34 +1,30 @@
 from abc import ABC
+from typing import Any, ClassVar, Dict, Optional, Type, Union
+from pydantic import validator
 
-from typing import Optional
-
-from ..message import InitRequestResponseReason
+from ..scheduler_request import SchedulerRequestResponse, UNSUCCESSFUL_JOB
+from ..message import AbstractInitRequest, InitRequestResponseReason
 from .external_request_response import ExternalRequestResponse
 from .model_exec_request import ModelExecRequest
+from .model_exec_request_response_body import ModelExecRequestResponseBody
 
 
 class ModelExecRequestResponse(ExternalRequestResponse, ABC):
 
-    _data_dict_key_job_id = "job_id"
-    _data_dict_key_output_data_id = "output_data_id"
-    _data_dict_key_scheduler_response = "scheduler_response"
-    response_to_type = ModelExecRequest
+    response_to_type: ClassVar[Type[AbstractInitRequest]] = ModelExecRequest
     """ The type of :class:`AbstractInitRequest` for which this type is the response"""
 
-    @classmethod
-    def _convert_scheduler_response_to_data_attribute(cls, scheduler_response=None):
-        if scheduler_response is None:
-            return None
-        elif isinstance(scheduler_response, dict) and len(scheduler_response) == 0:
-            return {}
-        elif isinstance(scheduler_response, dict):
-            return scheduler_response
-        else:
-            return {
-                cls._data_dict_key_job_id: scheduler_response.job_id,
-                cls._data_dict_key_output_data_id: scheduler_response.output_data_id,
-                cls._data_dict_key_scheduler_response: scheduler_response.to_dict(),
-            }
+    data: Optional[Union[ModelExecRequestResponseBody, Dict[str, Any]]] = None
+
+    @validator("data", pre=True)
+    def _convert_data_field(cls, value: Optional[Union[SchedulerRequestResponse, ModelExecRequestResponseBody, Dict[str, Any]]]) -> Optional[Union[ModelExecRequestResponseBody, Dict[str, Any]]]:
```

Comment:
Yeah... the type hints are certainly getting a little unwieldy. Which do you find more readable?

```python
# 1.
Optional[Union[SchedulerRequestResponse, ModelExecRequestResponseBody, Dict[str, Any]]]) -> Optional[Union[ModelExecRequestResponseBody, Dict[str, Any]]]

# 2.
Union[SchedulerRequestResponse, ModelExecRequestResponseBody, Dict[str, Any], None]) -> Union[ModelExecRequestResponseBody, Dict[str, Any], None]:
```

Reply:
In later versions of python this will look nicer:

```python
SchedulerRequestResponse | ModelExecRequestResponseBody | Dict[str, Any] | None -> ModelExecRequestResponseBody | Dict[str, Any] | None:
```

Reply:
I was thinking more along the lines of:

```python
MODEL_EXEC_DATA_FIELD = Union[
    SchedulerRequestResponse,
    ModelExecRequestResponseBody,
    Dict[str, Any]
]

MODEL_EXEC_RESPONSE_DATA = Union[ModelExecRequestResponseBody, Dict[str, Any]]
```

and later defining:

```python
def _convert_data_field(cls, value: MODEL_EXEC_DATA_FIELD = None) -> Optional[MODEL_EXEC_RESPONSE_DATA]:
```

This isn't something you need to move on if you don't think it's worth it. It's just a suggestion for something I found a tad confusing.

Reply:
> This isn't something you need to move on if you don't think it's worth it. It's just a suggestion for something I found a tad confusing.

Yeah that is fair. This is an unfortunate artifact of the previous implementation's invariants clashing with pydantic.

Reply:
So, looking closer at this, I think we could easily add an alias to simplify _part_ of things:

``` Python
data: Optional[Union[ModelExecRequestResponseBody, Dict[str, Any]]] = None

RESPONSE_DATA_SOURCE = Union[SchedulerRequestResponse, ModelExecRequestBody, Dict[str, Any], None]

def _convert_data_field(cls, value: RESPONSE_DATA_SOURCE) -> Optional[Union[ModelExecRequestResponseBody, Dict[str, Any]]]:
```

I get the impression (but am not totally sure; can you confirm @aaraney?) that we can't do something similar and apply it to the `data` type hint for reasons related to Pydantic.  

I suppose we could do this though:

``` Python
# Note that this field's type hint and the "RESPONSE_DATA" alias defined below must be kept in sync,
# but that Pydantic will not recognize the alias properly if used here
data: Optional[Union[ModelExecRequestResponseBody, Dict[str, Any]]] = None

# Note that this alias and the above "data" field's type hint must be kept in sync
RESPONSE_DATA = Optional[Union[ModelExecRequestResponseBody, Dict[str, Any]]]
RESPONSE_DATA_SOURCE = Union[SchedulerRequestResponse, ModelExecRequestBody, Dict[str, Any], None]

def _convert_data_field(cls, value: RESPONSE_DATA_SOURCE) -> RESPONSE_DATA:
```

Thoughts @aaraney and @christophertubbs?

Reply:
@robertbartel, the `data` field's type hint can be specified using a type alias. Ill add the type aliases mentioned above to improve readability.